### PR TITLE
Installer fix

### PR DESCRIPTION
**Pull request by craigh**
| Q                 | A
| ----------------- | ---
| Bug fix?          | yes
| New feature?      | no
| BC breaks?        | maybe?
| Deprecations?     | no
| Refs tickets      | #3962
| License           | MIT
| Changelog updated | not yet

## Description
trying to get project to build - e.g. composer will full work and run all scripts
also make sure that console runs properly

## Todos
- [x] mmenu is currently removed and must be re-added (probably refs #3982)
- [x] Liip/ImagineBundle is currently removed (error: Root image path not resolvable)
- [x] session.storage override in console is disabled refs 932f563fdaa9ae7b54289cc858f82a1bf1c7d267
- [x] probably other stuff.

### CODE REVIEW

**Comment by craigh** [comment]
@Guite - your feedback is welcome. I will continue to work on this branch as I find time. You're assistance with the ImagineBundle would be very helpful. 

Of course, there are multiple things wrong here, this is only a work in progress.

File: src/lib/Zikula/Bundle/CoreInstallerBundle/Util/ZikulaRequirements.php [hunk 1]
```diff
@@ -13,9 +13,9 @@
 
 namespace Zikula\Bundle\CoreInstallerBundle\Util;
 
-use Requirement;
+use Symfony\Requirements\Requirement;
 use Symfony\Component\Intl\Exception\MethodArgumentValueNotImplementedException;
```
Comment (Guite) [inline src/lib/Zikula/Bundle/CoreInstallerBundle/Util/ZikulaRequirements.php]: Please sort imports alphabetically.

File: src/lib/Zikula/Bundle/CoreInstallerBundle/composer.json [hunk 2]
```diff
@@ -14,9 +14,17 @@
...
         "symfony/framework-bundle": "4.*",
+        "symfony/console": "^4.3",
+        "symfony/requirements-checker": "dev-patch-1 as 1.*",
         "matthiasnoback/symfony-console-form": "3.*",
         "zikula/wizard": "3.*"
     },
+    "repositories": [
+        {
+            "type": "git",
+            "url": "https://github.com/craigh/requirements-checker"
```
Comment (Guite) [inline src/lib/Zikula/Bundle/CoreInstallerBundle/composer.json]: What is the reason for using your fork? You did no changes there so far.

Comment (Guite) [reply src/lib/Zikula/Bundle/CoreInstallerBundle/composer.json]: Found it: https://github.com/craigh/requirements-checker/commit/7149dcfdac06eec8c6f7efc0b1e2ecb4e0d25261

File: src/lib/Zikula/Core/Controller/AbstractController.php [hunk 3]
```diff
@@ -108,7 +108,7 @@ protected function decorateTranslator(array $parameters): array
      * Returns a NotFoundHttpException; this will result in a 404 response code.
      * Usage example: throw $this->createNotFoundException();
      */
-    public function createNotFoundException(string $message = 'Not Found', ?Throwable $previous = null): NotFoundHttpException
+    public function createNotFoundException(string $message = 'Access Denied.', \Throwable $previous = null): NotFoundHttpException
```
Comment (Guite) [inline src/lib/Zikula/Core/Controller/AbstractController.php]: Please change `Access Denied.` back to `Not Found` like shown at https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php#L291

File: src/system/ThemeModule/EventListener/DefaultPageAssetSetterListener.php [hunk 4]
```diff
@@ -68,7 +68,7 @@ public function __construct(
         string $bootstrapJavascriptPath,
         string $bootstrapFontAwesomeStylesheetPath,
         string $fontAwesomePath,
-        string $bootstrapStylesheetPath
+        $bootstrapStylesheetPath // cannot cast to bool because set with expression language
```
Comment (Guite) [inline src/system/ThemeModule/EventListener/DefaultPageAssetSetterListener.php]: Please replace `bool` by `string`.

**Comment by Guite** [comment]
I like the approach: getting the composer instance from the event seems a nice to migrate our custom actions.

**Comment by craigh** [comment]
@Guite can you help with this a bit? A lot of it is working now. When in `dev` mode, If you do `bin/console z:i:s` and complete it, then `bin/console z:i:f` you will see that there is a memory error when it is trying to install the Theme module. I can't figure it out.

Also, perhaps you can figure out the problem with ImagineBundle. I could not get the configuration correct (error was `Root image path not resolvable`). Since I don't use Imagine, I don't even know where to begin. I tried something like `data_root: '%kernel.root_dir%/../web/'` with no success.

**Comment by Guite** [comment]
I've added a few fixes.

> memory error when it is trying to install the Theme module
disable `dev` mode, the memory problem is probably caused by the profiler; or set `memory_limit=-1` for testing

one guess is that the theme table contains fields named `type` and `system` which could make problems being a reserved word...

```
Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'system SMALLINT NOT NULL, state SMALLINT NOT NULL, xhtml SMALLINT NOT NULL, PRIM' at line 1' while executing DDL: CREATE TABLE themes (id INT AUTO_INCREMENT NOT NULL, name VARCHAR(64) NOT NULL, type SMALLINT NOT NULL, displayname VARCHAR(64) NOT NULL, description VARCHAR(255) NOT NULL, version VARCHAR(10) NOT NULL, contact VARCHAR(255) NOT NULL, admin SMALLINT NOT NULL, user SMALLINT NOT NULL, system SMALLINT NOT NULL, state SMALLINT NOT NULL, xhtml SMALLINT NOT NULL, PRIMARY KEY(id)) DEFAULT CHARACTER SET UTF8 COLLATE `UTF8_unicode_ci` ENGINE = InnoDB
```

Also the extensions table contains a field named `type`.

Maybe worth renaming two fields to `extensionType`, `themeType` and remove `system` entirely? slightly refs #3644


Also at the very end of the CLI install finish process I get:

```
Finalize parameters
 Finalizing parameters

In getThemeEntityRepositoryService.php line 3:
                                                              
  Maximum function nesting level of '256' reached, aborting!
```

**Comment by Guite** [comment]
> I've added a few fixes.

Sorry, forgot to push. Will do later.

**Comment by craigh** [comment]
✅ I am using prod mode apparently (thought it was dev until I checked)
❌ Still getting same 'out of memory' errors despite most recent commit that fixes reserved word issue.

I'm guessing the `Maximum function nesting level of '256' reached, aborting!` is the real problem. Somehow we must have a memory leak or much deeper nesting of objects than I thought. Do you have other ideas?

**Comment by craigh** [comment]
@Guite can you please take a look at this? I've been staring at it for hours and not coming up with a solution.

Basically, the problem is that Twig is not registering any custom extensions. The services are present and properly autowired/configured and include `twig.extension` tag as needed. for example `bin/console debug:container ModVarExtension` produces expected result. but `bin/console debug:twig` will not show the functions registered in said extension. And `bin/console lint:twig` shows that all the custom functions/filters/etc are not recognized.

I'm hoping I missed something obvious. thanks!

**Comment by Guite** [comment]
> Basically, the problem is that Twig is not registering any custom extensions.

Do you refer to the decorated events? I experienced similar problems before when I worked on #2221 in last spring - you might read through all links referenced in that ticket. But I also think this is outside of the scope of this PR.

**Comment by craigh** [comment]
> Do you refer to the decorated events?

no. I simply mean that Twig cannot find custom extensions. Look at the errors in the travis build, you will see what I mean.

> But I also think this is outside of the scope of this PR.

The scope of the PR is to get the project to build without error. It is also to get both the CLI and web installer to fully function producing at least a first page error free.

**Comment by Guite** [comment]
Okay. I thought you wanted to include #2221 entirely.

> I simply mean that Twig cannot find custom extensions.

Auto configuration of Twig-related service tags is done in the TwigBundle:
https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php#L144

Is this bundle not available anymore?

Then we could do this in one of our bundles.
Examples:
- https://github.com/zikula/core/blob/master/src/lib/Zikula/Bundle/CoreBundle/DependencyInjection/CoreExtension.php#L42
- https://github.com/zikula/core/blob/master/src/lib/Zikula/Bundle/HookBundle/DependencyInjection/ZikulaHookExtension.php#L33
- https://github.com/zikula/core/blob/master/src/system/UsersModule/DependencyInjection/ZikulaUsersExtension.php#L31

I would recommend doing this in the Theme module then.

**Comment by craigh** [comment]
we shouldn't need to do this. The TwigBundle is present and some services from it are in use. you can use `bin/console debug:container` to see what is there and registered.

**Comment by craigh** [comment]
yes #2221 will need to be fully included here I think

File: src/system/ThemeModule/Resources/config/services.yml [hunk 5]
```diff
@@ -84,10 +95,11 @@ services:
           $combine: '%zikula_asset_manager.combine%'
 
     Zikula\ThemeModule\Bridge\Twig\EventEnabledTwigEngine:
-        decorates: templating.engine.twig
+        decorates: twig
         arguments:
-          $parser: '@templating.name_parser'
-          $locator: '@templating.locator'
+          $loader: '@twig.loader'
```
Comment (craigh) [inline src/system/ThemeModule/Resources/config/services.yml]: @Guite given your previous comments, perhpahs this is the problem. Should `$loader` be something else?

Comment (Guite) [reply src/system/ThemeModule/Resources/config/services.yml]: No, I think the missing extensions are not related to any type of template loader.

File: src/system/RoutesModule/AppSettings.php [hunk 6]
```diff
@@ -1,17 +1,17 @@
 <?php
 
-declare(strict_types=1);
```
Comment (craigh) [inline src/system/RoutesModule/AppSettings.php]: in the rest of the core, the declaration is at the top (like this) not after the comment, like below. Why change it?

Comment (Guite) [reply src/system/RoutesModule/AppSettings.php]: Demanded by phpcs / PHP_CodeSniffer. Maybe to conform to PSR-12 standard.

Also it we are actually documenting the file, not the `namespace` keyword. Thus, the comment must come first. See also https://github.com/squizlabs/PHP_CodeSniffer/issues/1800#issuecomment-352634571

You don't need to mind about it. I will change rest of the core later. Refs #3987

**Comment by craigh** [comment]
Screen Shot 2019-12-14 at 9 19 45 AM

**Comment by craigh** [comment]
regarding 838fa1c4407fb0fcaaf3b033dcc425608ab97784

for reference:
Screen Shot 2019-12-14 at 8 48 38 AM

paths can be tricky. especially because different production environments can be in myriad sub-directories.

**Pull request merged**